archive/tar.go

```diff
@@ -648,3 +649,22 @@ func copyBuffered(ctx context.Context, dst io.Writer, src io.Reader) (written in
 	return written, err
 
 }
+
+// rootPathForHardlink returns target linkname, evaluating and bounding any
+// symlink to the parent directory.
+//
+// NOTE: Allow hardlink to the softlink, not the real one. For example,
+//
+//	touch /tmp/zzz
+//	ln -s /tmp/zzz /tmp/xxx
+//	ln /tmp/xxx /tmp/yyy
+//
+// /tmp/yyy should be softlink which be same of /tmp/xxx, not /tmp/zzz.
+func rootPathForHardlink(root, linkname string) (string, error) {
+	ppath, base := filepath.Split(linkname)
+	ppath, err := fs.RootPath(root, ppath)
+	if err != nil {
+		return "", err
+	}
+	return filepath.Join(ppath, base), nil
```

Comment:
The result of this should still be checked before returning since base could be `".."` and this would return the parent directory. The link operation should fail but it should not get all the way to making a syscall with the parent directory as an arg.